x/ibc/07-tendermint/types/proposal_handle.go

```diff
@@ -40,33 +40,49 @@ func (cs ClientState) CheckProposedHeaderAndUpdateState(
 		)
 	}
 
-	if cs.IsFrozen() {
+	switch {
+
+	case cs.IsFrozen():
 		if !cs.AllowUpdateAfterMisbehaviour {
 			return nil, nil, sdkerrors.Wrap(clienttypes.ErrUpdateClientFailed, "client is not allowed to be unfrozen")
 		}
 
 		// unfreeze the client
 		cs.FrozenHeight = clienttypes.Height{}
 
-		// if the client is frozen but not expired, do full validation of the header
-		// NOTE: the client may be frozen again since the misbehaviour evidence may
-		// not be expired yet
-		if !cs.IsExpired(consensusState.Timestamp, ctx.BlockTime()) {
+		// if the client is expired we unexpire the client using softer validation, otherwise
+		// full validation on the header is performed.
+		if cs.IsExpired(consensusState.Timestamp, ctx.BlockTime()) {
+			return cs.unexpireClient(consensusState, tmHeader, ctx.BlockTime())
+		} else {
+			// NOTE: the client may be frozen again since the misbehaviour evidence may
+			// not be expired yet
 			return cs.CheckHeaderAndUpdateState(ctx, cdc, clientStore, header)
 		}
```

Comment:
`return cs.unfreezeClient(...)`?

Reply:
I think it's fine for now. `unexpire` is needed because it is called twice